### PR TITLE
created class RegistrationException;

### PR DESCRIPTION
**Pull request by aksoli666**
implements class RegistrationServiceImplTest;
override method register(User user);

### CODE REVIEW

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+        validUser.setPassword("validPassword");
+        validUser.setAge(20);
+
+        registrationServiceImpl.register(validUser);
+
+        assertNotNull(storageDao.get("ValidUser"));
+    }
+
+    @Test
+    public void register_differentLogin_ok() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:42]: move all ok-cases into one ok-method

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        user.setLogin("abc");
+        user.setPassword("password");
+        user.setAge(20);
+
+        assertThrows(RegistrationException.class, () -> registrationServiceImpl.register(user));
+    }
+
+    @Test
+    public void register_lengthLoginFive_notOk() {
+        User user = new User();
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:110]: combine to register_invalidLogin_notOk

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+        user.setLogin("UserAgeNegative6");
+        user.setPassword("password");
+        user.setAge(-6);
+
+        assertThrows(RegistrationException.class, () -> registrationServiceImpl.register(user));
+    }
+
+    @Test
+    public void register_negativeAgeEighteen_notOk() {
+        User user = new User();
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:240]: register_invalidAge_notOk

**Review by Rommelua — CHANGES_REQUESTED** [review]
good job

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 4]
```diff
@@ -5,10 +5,33 @@
...
+            throw new RegistrationException();
+        }
+
+        if (user.getAge() == null || user.getAge() < MIN_AGE) {
+            throw new RegistrationException();
+        }
+
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new RegistrationException();
+        }
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:33]: Add informative message for every throwed exception

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+import core.basesyntax.dao.StorageDao;
+import core.basesyntax.dao.StorageDaoImpl;
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class RegistrationServiceImplTest {
+    private final RegistrationServiceImpl registrationServiceImpl = new RegistrationServiceImpl();
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:15]: init this field in method BeforeAll or BeforeEach

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        assertTrue(Storage.people.add(validUser));
+
+        validUser.setAge(25);
+        assertTrue(Storage.people.add(validUser));
+
+        validUser.setAge(60);
+        assertTrue(Storage.people.add(validUser));
+
+        validUser.setAge(Integer.MAX_VALUE);
+        assertTrue(Storage.people.add(validUser));
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:86]: Don`t reuse one user for different age

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    @Test
+    public void register_invalidLogin_notOk() {
+        User firstUser = new User();
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:91]: non informative naming of user, better name it like firstValidUser or firstInvalidUser, userInvalidAge etc

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        assertThrows(RegistrationException.class, () -> registrationServiceImpl.register(user),
+                "Username length 0. Login must be at least 6 characters long");
+
+        user.setLogin("abc");
+        assertThrows(RegistrationException.class, () -> registrationServiceImpl.register(user),
+                "Username length 3. Login must be at least 6 characters long");
+
+        user.setLogin("abcde");
+        assertThrows(RegistrationException.class, () -> registrationServiceImpl.register(user),
+                "Username length 5. Login must be at least 6 characters long");
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:123]: In one user only one assert for login, don`t reuse it

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        assertThrows(RegistrationException.class, () -> registrationServiceImpl.register(user),
+                "Age: 6. User must be at least 18 years old");
+
+        user.setAge(17);
+        assertThrows(RegistrationException.class, () -> registrationServiceImpl.register(user),
+                "Age: 17. User must be at least 18 years old");
+
+        user.setAge(Integer.MIN_VALUE);
+        assertThrows(RegistrationException.class, () -> registrationServiceImpl.register(user),
+                "User must be at least 18 years old");
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:183]: Same situation

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Review by sokimaaa — APPROVED** [review]

**Pull request closed without merging**